src/app/clusters/scenes/SceneTableImpl.cpp

```diff
@@ -0,0 +1,619 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/clusters/scenes/SceneTableImpl.h>
+#include <lib/support/DefaultStorageKeyAllocator.h>
+#include <stdlib.h>
+
+namespace chip {
+namespace scenes {
+
+enum SceneImplTLVTag
+{
+    kTagSceneCount = 1,
+    kTagNext,
+};
+
+using SceneTableEntry = DefaultSceneTableImpl::SceneTableEntry;
+using SceneStorageId  = DefaultSceneTableImpl::SceneStorageId;
+using SceneData       = DefaultSceneTableImpl::SceneData;
+
+struct FabricHavingSceneList : public CommonPersistentData::FabricList
+{
+    CHIP_ERROR UpdateKey(StorageKeyName & key) override
+    {
+        key = DefaultStorageKeyAllocator::SceneFabricList();
```

Comment:
This is very risky to reuse and assume the system implementation of GroupDataProviderImpl is what is used, since you share data structures. Is the coupling safe?

Reply:
The coupling was tested on device to the extent that the presence of the SceneTable does not break the GroupDataProvider and that the absence of the GroupDataProvider does not break the SceneTable. Comments and verifications will be added regarding the dependency on the GroupDataProvider implementation. I could also switch to a different key at cost of duplicating data in the flash.